core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java

```diff
@@ -0,0 +1,135 @@
+ *       agg_right <- aggRight(mask = agg_mask_right)
+ *     project(agg_mask_left = filter_left, agg_mask_left = filter_right)
+ *       filter(filter_left or filter_right)
+ *         table scan
+ * </pre>
+ * <p>
+ * This makes the plan to read the source table only once,
+ * resulting in the potentially significant performance improvement.
+ */
+public class FuseCrossJoinedGlobalAggregations
+        implements Rule<JoinNode>
+{
+    private static final Capture<AggregationNode> LEFT_AGGREGATION_NODE = newCapture();
+    private static final Capture<AggregationNode> RIGHT_AGGREGATION_NODE = newCapture();
+    private static final Pattern<JoinNode> PATTERN = join()
+            .matching(JoinNode::isCrossJoin)
+            .with(left()
+                    .matching(aggregation()
+                            .matching(AggregationNode::hasSingleGlobalAggregation)
+                            .capturedAs(LEFT_AGGREGATION_NODE)))
+            .with(right()
+                    .matching(aggregation()
+                            .matching(AggregationNode::hasSingleGlobalAggregation)
+                            .capturedAs(RIGHT_AGGREGATION_NODE)));
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isCombineSimilarSubPlans(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        AggregationNode left = captures.get(LEFT_AGGREGATION_NODE);
+        AggregationNode right = captures.get(RIGHT_AGGREGATION_NODE);
+
+        return new PlanNodeFuser(context).fuse(left, right)
+                .map(fused -> {
+                    checkArgument(
+                            fused.leftFilter().equals(TRUE_LITERAL) && fused.rightFilter().equals(TRUE_LITERAL),
+                            "Expected both fused filter to be TRUE but got left %s, right %s",
+                            fused.leftFilter(),
+                            fused.rightFilter());
+                    PlanNode result = fused.plan();
+
+                    // verify if some symbol need to be aliased in the result
+                    Map<Symbol, SymbolReference> additionalProjections = new HashMap<>();
+                    Set<Symbol> resultOutputSymbols = ImmutableSet.copyOf(result.getOutputSymbols());
+                    for (Symbol originalOutputSymbol : node.getOutputSymbols()) {
+                        if (!resultOutputSymbols.contains(originalOutputSymbol)) {
+                            additionalProjections.put(originalOutputSymbol, fused.symbolMapping().requiredMap(originalOutputSymbol).toSymbolReference());
+                        }
+                    }
+                    if (!additionalProjections.isEmpty()) {
+                        result = new ProjectNode(
+                                context.getIdAllocator().getNextId(),
```

Comment:
We can probably use `node.getId()` here as the original JoinNode should get discarded